src/main/java/com/google/devtools/build/lib/analysis/CommandHelper.java

```diff
@@ -282,7 +297,7 @@ public String expandLabelsHeuristically(String expr) {
       RuleContext ruleContext, String command, CommandConstructor constructor) {
     List<String> argv;
     Artifact scriptFileArtifact = null;
-    if (command.length() <= maxCommandLength) {
+    if (command.length() <= getMaxCommandLength(ruleContext.getHostConfiguration().getOS())) {
```

Comment:
Most of the uses you have seem to be using it with `ruleContext.getHostConfiguration().getOS()`. This is a bit inflexible given the presence of [execution groups](https://docs.bazel.build/versions/3.5.0/exec-groups.html), but we do have a way to tell what platform is being used as the execution platform:
```
ruleContext.getToolchainContext(execGroupName).executionPlatform();
```

You can then call `executionPlatform.constraints().hasConstraintValue()` to see whether the current execution platform has the Windows OS constraint (you'll need an instance of `ConstraintValueInfo`, or we could add an overload that just takes a Label and ignores the configuration).

This could probably all be packaged up more hygenically, but the code is in place.

See the docs for [`RuleContext.getToolchainContext()`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java;bpv=1;bpt=1;l=1208?ss=bazel&q=RuleContext&gsn=getToolchainContext&gs=kythe%3A%2F%2Fgithub.com%2Fbazelbuild%2Fbazel%3Flang%3Djava%3Fpath%3Dcom.google.devtools.build.lib.analysis.RuleContext%23b5d9b148eb20de8be0d1443d570c851cc8d3eae5961072a9ff0cf471ed53d612) (and the various related methods), and for [`ToolchainContext.executionPlatform()`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/ToolchainContext.java;drc=49c9475aa7db252c19157ee059c40862a412df5d;bpv=1;bpt=1;l=28?gsn=executionPlatform&gs=kythe%3A%2F%2Fgithub.com%2Fbazelbuild%2Fbazel%3Flang%3Djava%3Fpath%3Dcom.google.devtools.build.lib.analysis.ToolchainContext%232330a5d89a3cbbc320688c988fe74955ca37910f80c897258b25e514ae69d181).

Reply:
What's the relationship between constraints and execution properties?

Instead of using OS, I could check the `OSFamily` setting:
https://github.com/bazelbuild/remote-apis/blob/master/build/bazel/remote/execution/v2/platform.md

That would require adding that to the execution properties by default - otherwise local execution wouldn't work, I think.

If using the execution properties isn't desirable, any suggestions for how to get a constraint value? It would have to be a universal label that all users of Bazel agree upon.

Reply:
How do you feel about adding a method to Platform that returns an OS instance, as a temporary measure?

Reply:
Constraints and execution properties are entirely separate: ideally Bazel never even looks inside the execution properties, just hands them off to remote execution. Currently only remote platforms set them, and also the properties defined by the RE API are only valid for platforms intended to be used for remote execution: when we start using platforms to define the properties for sandboxing, for example, they wouldn't make much sense (or we'd be using remote execution API for non-remote execution purposes).

We have several useful constraints defined in [`@bazel_tools//platforms`](https://cs.opensource.google/bazel/bazel/+/master:tools/platforms/BUILD.tools), including OS and CPU constraints. These are actually just aliases for the values in https://github.com/bazelbuild/platforms, which are available to all builds (or else the aliases wouldn't work), so those would be fine to use.

As a shortcut, I'm okay with adding a PlatformInfo.getOS method, but how would we set it? Just add code in PlatformInfo that knows which constraints mean windows, darwin, etc? Is this actually easier than having code that cares about Windows ask if the PlatformInfo has the windows constraint?

Reply:
I don't see exec properties as being purely for remote execution. For example, exec properties could also be used to configure the sandbox whether it should allow networking or not, or which local paths to map into the sandbox (currently this is only possible through per-build settings, which is not ideal).

For `OSFamily` specifically, I think the constraints and `OSFamily` setting should be consistent, and, ideally, this would be ~enforced by Bazel (for some value of enforced).

Regardless, I'm a bit at a loss as to how I can get the right object into - say - the `TestStrategy` to check for a Windows constraint.

Reply:
Execution properties as a general concept make sense for all platforms. The specific definitions of the keys and values are dependent on the specific execution strategy being used. Especially since Bazel's remote execution and Google's internal remote execution use different keys, I don't want to tie Bazel into using one specific set.

The constraints matching the execution properties has to be managed by whoever defines the platforms: if Bazel checks at all, that should be inside the remote execution strategy (which is what defines the keys and values), and not generic to all execution strategies (because they could all use different keys and values).

The [`ConstraintCollection`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/platform/ConstraintCollection.java) class currently defines `get(ConstraintSettingInfo)` and `hasConstraintValue(ConstraintValueInfo)` methods which take providers, but those providers are fairly lightweight. The [`ConstraintSettingInfo`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/platform/ConstraintSettingInfo.java) is a wrapper around a Label defining the setting (and a default constraint that could be used). [`ConstraintValueInfo`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/platform/ConstraintValueInfo.java) is just a `ConstraintSettingInfo` and a label defining the value.

The way we get these for toolchain resolution is to actually fetch the ConfiguredTarget and extract the providers, which is heavyweight for your purposes. Since the OS constraint doesn't provide a default, I am fine with adding an overload of `ConstraintCollection.hasConstraintValue` which takes a Label of a constraint value, and checks to see if any directly declared constraint value matches. This will will any constraint values that are present due to being the default, but this will work for your purposes. Then you can call like this:
```
boolean windows = getExecutionPlatform().constraints().hasConstraintValue(Label.parseAbsoluteUnchecked("@platforms//os:windows");
```

Would that work for you? Adding the overload should be straightforward.